### PR TITLE
[SPARK-38617][SQL][WEBUI] Show Spark rule and phase timings in SQL UI and REST API

### PR DESCRIPTION
**Pull request by yliou**
### What changes were proposed in this pull request?
The changes proposed are to add a clickable field on SQL UI to show timing of spark phases and rule timing information from QueryPlanningTracker along with a separate REST endpoint to show the timing statistics. Adds a SQLConf to control the number of rules that 
show up on the UI and the REST endpoint call (default value is 15).
UI Change: Screen Shot 2022-05-25 at 11 04 50 AM
New REST API endpoint output: 
[SPARK-38617RestOutput.txt](https://github.com/apache/spark/files/8773891/SPARK-38617RestOutput.txt)

### Why are the changes needed?
Currently user is unable to see Spark phase timing information or which Spark rules take the most time for a specific query execution on the UI. This PR adds that information.

### Does this PR introduce _any_ user-facing change?
Yes on the UI

### How was this patch tested?
Tested locally and added a class to test the new endpoint.

### CODE REVIEW

**Comment by martin-g** [comment]
[SPARK38617.txt](https://github.com/apache/spark/files/8326833/SPARK38617.txt) is in rich text format. It would be easier to review if it was plain text.

File: sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    val salary: DataFrame =
+      spark.sparkContext.parallelize(
+        Salary(0, 2000.0) ::
+          Salary(1, 1000.0) :: Nil).toDF()
+
+    val salaryDF = salary.withColumnRenamed("personId", "id")
+    val ds = person.join(salaryDF, "id")
+      .groupBy("name", "age", "salary").avg("age", "salary")
+      .filter(_.getAs[Int]("age") <= 30)
```
Comment (martin-g) [inline sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala:114]: the filtering filters nothing with the provided two persons. Is this intentional ?

**Comment by ulysses-you** [comment]
https://github.com/apache/spark/pull/35856 has tried the similar things although at present it miss the details of the compile time for each phase. And its scope also includes the AQE, I think is good to go.

File: sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala (hunk 1 above)
Comment (yliou) [reply sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala:114]: Yes

File: sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlCompilerResource.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+private[v1] class SqlCompilerResource extends BaseAppResource with Logging {
+
+  @GET
+  def compilerStat(
+                    @DefaultValue("0") @QueryParam("offset") offset: Int,
+                    @DefaultValue("20") @QueryParam("length") length: Int): Seq[CompileData] = {
+    withUI { ui =>
+      val sqlStore = new SQLAppStatusStore(ui.store.store)
+      val appid = ui.store.environmentInfo().sparkProperties.toMap.getOrElse("spark.app.id", "")
+      sqlStore.executionsList(offset, length).map { exec =>
```
[the remark this answers is not in the document]
Comment (yliou) [reply sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlCompilerResource.scala:39]: It shouldn't be needed. I added SqlCompilerResource.scala based on SqlResource.scala

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala [hunk 3]
```diff
@@ -79,6 +80,14 @@ class SQLAppStatusStore(
   def planGraph(executionId: Long): SparkPlanGraph = {
     store.read(classOf[SparkPlanGraphWrapper], executionId).toSparkPlanGraph()
   }
+
+  def getCompilerStats(executionId: Long): QueryPlanningTracker = {
+    try {
+      store.read(classOf[CompilerStats], executionId).queryStats
+    } catch {
+      case _: NoSuchElementException => new QueryPlanningTracker()
```
[the remark this answers is not in the document]
Comment (yliou) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusStore.scala:88]: I added this try catch statement as a precaution after adding the if statement on SQLAppStatusListener.scala:393. Some unit tests were failing without the if statement. I can remove the try catch statement if desired.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala [hunk 4]
```diff
@@ -160,4 +162,23 @@ class QueryPlanningTracker {
     }
   }
 
+  override def toString(): String = {
```
[the remark this answers is not in the document]
Comment (yliou) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: Renamed method to compileStatsString

**Comment by yliou** [comment]
> [SPARK38617.txt](https://github.com/apache/spark/files/8326833/SPARK38617.txt) is in rich text format. It would be easier to review if it was plain text.

Attached new file with plain text format.

**Comment by yliou** [comment]
> #35856 has tried the similar things although at present it miss the details of the compile time for each phase. And its scope also includes the AQE, I think is good to go.

What does good to go mean here as I don't see an approval at #35856? From what I see the changes at #35856 are different and complimentary with this PR as that PR adds information regarding AQE plan description changes and this PR has timing of spark phases and rule timing information.

**Comment by sigmod** [comment]
cc @jchen5

**Comment by ulysses-you** [comment]
@yliou  I agree both can help for troubleshooting. And I have mentioned the difference, `it miss the details of the compile time for each phase`. That PR introduces a disk-based store so I'm thinking if can re-use the same code path. IMO it would be better to go with that PR first at least.

**Comment by mridulm** [comment]
+CC @shardulm94, @thejdeep

**Comment by sigmod** [comment]
Can we use a more meaning JSON field names? E.g.,
`value` -> `timeMs`
`name` -> `phase`

 ```
{
    “name” : “optimization”,
    “value” : “373"
  }, {
    “name” : “analysis”,
    “value” : “12"
  } ],
  “ruleInfo” : [ {
    “ruleName” : “org.apache.spark.sql.catalyst.optimizer.ConvertToLocalRelation”,
    “value” : “32.659566”,
    “numInvocations” : 5,
    “numEffectiveInvocations” : 3
  }, 
  ....
]

```

**Comment by yliou** [comment]
@ulysses-you got it. I think it is possible to re-use the same code path but not sure how that'd work with ExecutionPage.scala to get the info on the UI.

**Comment by yliou** [comment]
> Can we use a more meaningful JSON field names? E.g., `value` -> `timeMs` `name` -> `phase`
> 
> ```
> {
>    “phase” : “optimization”,
>    “timeMs” : “373"
>  }, {
>    “phase” : “analysis”,
>    “timeMs” : “12"
>  } ],
>  “ruleInfo” : [ {
>    “ruleName” : “org.apache.spark.sql.catalyst.optimizer.ConvertToLocalRelation”,
>    “timeMs” : “32.659566”,
>    “numInvocations” : 5,
>    “numEffectiveInvocations” : 3
>  }, 
>  ....
> ]
> ```
> 
> Also, can `timeMs` be a number instead of a string?

Updated Json field names and changed `timeMs` to be a number

**Comment by linhongliu-db** [comment]
@ulysses-you @yliou adding debug information is always good. But I'm worried about two issues:
1. we should be careful when adding information to live UI's in-memory KV store. sooner or later, it will bring headaches
2. we should be careful when adding new REST API endpoints. since two PRs both for troubleshooting, maybe we can use the same endpoint.

File: sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.catalyst.plans.SQLHelper
+import org.apache.spark.sql.execution.metric.SQLMetricsTestUtils
+import org.apache.spark.sql.internal.SQLConf.ADAPTIVE_EXECUTION_ENABLED
+import org.apache.spark.sql.test.SharedSparkSession
+
+/**
+ * Sql Resource Public API Unit Tests running query and extracting the metrics.
+ */
+class SqlCompileResourceSuite
```
[the remark this answers is not in the document]
Comment (yliou) [reply sql/core/src/test/scala/org/apache/spark/status/api/v1/sql/SqlCompileResourceSuite.scala]: Thanks for catching the typo! I renamed the file.

**Comment by yliou** [comment]
Updated the PR to be more concise regarding the names of rules so that Spark event logs aren't as impacted as before in terms off size.

**Comment by yliou** [comment]
@dependabot reopen

**Comment by yliou** [comment]
Hi @ulysses-you @linhongliu-db now that https://github.com/apache/spark/pull/35856 got reverted and https://github.com/apache/spark/pull/38567 got merged, should I create another pull request for this feature to try to get it merged?

**Pull request closed without merging**